src/function/scalar/map/map.cpp

```diff
@@ -2,11 +2,77 @@
 #include "duckdb/common/string_util.hpp"
 #include "duckdb/parser/expression/bound_expression.hpp"
 #include "duckdb/function/scalar/nested_functions.hpp"
+#include "duckdb/function/list_aggregate_function.hpp"
+#include "duckdb/function/aggregate/nested_functions.hpp"
 #include "duckdb/common/types/data_chunk.hpp"
 #include "duckdb/common/pair.hpp"
 
 namespace duckdb {
 
+bool KeyListIsEmpty(list_entry_t *data, idx_t rows) {
+	for (idx_t i = 0; i < rows; i++) {
+		auto size = data[i].length;
+		if (size != 0) {
+			return false;
+		}
+	}
+	return true;
+}
+
+static void CheckKeyValidity(DataChunk &args) {
+	auto types = args.GetTypes();
+	if (types.empty()) {
+		return;
+	}
+	auto key_type = ListType::GetChildType(types[0]).id();
+	auto &array = args.data[0];
+	auto arg_data = FlatVector::GetData<list_entry_t>(array);
+	auto &entries = ListVector::GetEntry(array);
+	if (key_type == LogicalTypeId::SQLNULL) {
+		if (KeyListIsEmpty(arg_data, args.size())) {
+			return;
+		}
+		// The entire key list is NULL for one (or more) of the rows: (ARRAY[NULL, NULL, NULL])
+		throw InvalidInputException("Map keys can not be NULL");
+	}
+
+	VectorData list_data;
+	auto count = ListVector::GetListSize(array);
+	args.data[0].Orrify(args.size(), list_data);
+	auto validity = FlatVector::Validity(entries);
+	if (!validity.CheckAllValid(count)) {
+		throw InvalidInputException("Map keys can not be NULL");
+	}
+}
+
+static void CheckForKeyUniqueness(DataChunk &args, ExpressionState &state) {
+	// Create a copy of the arguments
+	auto types = args.GetTypes();
+	if (types.empty() || ListType::GetChildType(types[0]).id() == LogicalType::SQLNULL) {
+		return;
+	}
+
+	auto arg_data = FlatVector::GetData<list_entry_t>(args.data[0]);
+
+	DataChunk keys;
+	keys.Initialize(args.GetTypes());
+	args.Copy(keys);
+
+	// Split the copy to separate the keys
+	DataChunk remaining_columns;
+	keys.Split(remaining_columns, 1);
+
+	Vector unique_result(LogicalType::UBIGINT, args.size());
+	ListUniqueFunction(keys, state, unique_result);
+	for (idx_t i = 0; i < args.size(); i++) {
+		auto keys_length = arg_data[i].length;
+		auto unique_keys = FlatVector::GetValue<uint64_t>(unique_result, i);
+		if (unique_keys != keys_length) {
+			throw InvalidInputException("Map keys have to be unique!");
+		}
+	}
+}
+
```

Comment:
Nitpicking, wouldn't it be better if these are 
AreKeysUnique and AreKeysValid functions that return a bool with the exception being thrown in the MapFunction?
Also, how are these functions called when you have a map type not created by SQL?
e.g., if you have a malformed pyarrow table with a map type and then consume it  through duckdb, I think these functions won't be called, am I right?

Reply:
Hmm that could definitely be the case, I will test those cases

Reply:
```py
map_type = pa.map_(pa.int32(), pa.int32())
values = [
	[
		(3, 12),
		(3, 21)
	],
	[
		(5, 42)
	]
]
arrow_table = pa.table(
	{'detail': pa.array(values, map_type)}
)
rel = duckdb.from_arrow(arrow_table).fetchall()
```
```
[({'key': [3, 3], 'value': [12, 21]},), ({'key': [5], 'value': [42]},)]
```
Your hunch was correct, this error isn't caught

Reply:
Now the error is caught and tested for 👍